crunchers/cruncher_top_cards.h

```diff
@@ -0,0 +1,250 @@
+      top_cards_map_[card.rate].insert(e.cid);
+    }
+    events_list_.push_front(std::move(e));
+  }
+
+  void TimeWindowLeft(const CardEvent& e) {
+    const auto cit = cards_map_.find(e.cid);
+    if (cit != cards_map_.end()) {
+      auto& card = cit->second;
+      top_cards_map_[card.rate].erase(e.cid);
+      ApplyCardEvent(card, e.type, true /*rollback*/);
+      if (!card.Empty()) {
+        top_cards_map_[card.rate].insert(e.cid);
+      } else {
+        cards_map_.erase(e.cid);
+      }
+    }
+  }
+
+  void ApplyCardEvent(card_t& card, CTFO_EVENT event, bool rollback) {
+    int sign = rollback ? -1 : 1;
+    switch (event) {
+      case CTFO_EVENT::SEEN:
+        card.seen_count += sign;
+        break;
+      case CTFO_EVENT::SKIP:
+        card.skip_count += sign;
+        break;
+      case CTFO_EVENT::CTFO:
+        card.ctfo_count += sign;
+        break;
+      case CTFO_EVENT::TFU:
+        card.tfu_count += sign;
+        break;
+      case CTFO_EVENT::FAV_CARD:
+        card.fav_count += sign;
+        break;
+      case CTFO_EVENT::UNFAV_CARD:
+        card.unfav_count += sign;
+        break;
+      default:
+        return;
+    }
+    card.rate = args_.rate_calculator(
+        card.ctfo_count, card.skip_count, card.tfu_count, card.fav_count, card.unfav_count, card.seen_count);
+  }
+
+  void OnCardDeleted(const CardDeleted& e) {
+    const auto cit = cards_map_.find(e.key);
+    if (cit != cards_map_.end()) {
+      auto& top_bucket = top_cards_map_[cit->second.rate];
+      top_bucket.erase(e.key);
+      cards_map_.erase(cit);
+      if (top_bucket.empty()) {
+        top_cards_map_.erase(cit->second.rate);
+      }
+    }
+  }
+
+  void OnEventLogEntry(const EventLogEntry& e) {
+    current_us_ = e.server_us;
+    if (Exists<iOSGenericEvent>(e.event)) {
+      OnIOSGenericEvent(Value<iOSGenericEvent>(e.event));
+    }
+  }
+
+  void OnIOSGenericEvent(const iOSGenericEvent& e) {
+    auto& supported_events = current::ThreadLocalSingleton<std::map<std::string, CTFO_EVENT>>();
+    if (supported_events.empty()) {
+      supported_events.emplace("SEEN", CTFO_EVENT::SEEN);
+      supported_events.emplace("CTFO", CTFO_EVENT::CTFO);
+      supported_events.emplace("TFU", CTFO_EVENT::TFU);
+      supported_events.emplace("SKIP", CTFO_EVENT::SKIP);
+      supported_events.emplace("FAV", CTFO_EVENT::FAV_CARD);
+      supported_events.emplace("FAV_CARD", CTFO_EVENT::FAV_CARD);
+      supported_events.emplace("UNFAV", CTFO_EVENT::UNFAV_CARD);
+      supported_events.emplace("UNFAV_CARD", CTFO_EVENT::UNFAV_CARD);
+    }
+    try {
+      const CTFO_EVENT event = supported_events.at(e.event);
+      const std::string cid_str = e.fields.count("cid") ? e.fields.at("cid") : "";
+      TimeWindowsEnter(CardEvent{StringToCID(cid_str), event, current_us_});
+    } catch (const std::out_of_range&) {
+      // ignore unsupported events
+    }
+  }
+
+  CID StringToCID(const std::string& s) {
```

Comment:
`static`, a comment that it's a copy of the function from `../util.h`, refactor `../util.h` into two files one of which can be safely included from the outside, or plain `#include "../util.h"` if it works would be :+1: